common/djangoapps/third_party_auth/tests/specs/test_testshib.py

```diff
@@ -155,41 +203,154 @@ def test_configure_testshib_provider_with_cache_duration(self):
+                        'firstName': 'John',
+                        'lastname': 'Smith',
+                        'defaultFullName': 'John Smith',
+                        'email': 'john@smith.com',
+                    }
+                })
+            )
+
+        httpretty.register_uri(httpretty.GET, ODATA_USER_URL, content_type='application/json', body=user_callback)
+
+    def test_register_insufficient_sapsf_metadata(self):
+        """
+        Configure the provider such that it doesn't have enough details to contact the SAP
+        SuccessFactors API, and falls back to the data it receives from the SAML assertion.
+        """
+        self._configure_testshib_provider(
+            identity_provider_type='sap_success_factors',
+            other_settings='{"key_i_dont_need":"value_i_also_dont_need"}',
+        )
+        # Because we're getting details from the assertion, fall back to the initial set of details.
+        self.USER_EMAIL = "myself@testshib.org"
+        self.USER_NAME = "Me Myself And I"
+        self.USER_USERNAME = "myself"
+        super(SuccessFactorsIntegrationTest, self).test_register()
+
+    def test_register_sapsf_metadata_present(self):
+        """
+        Configure the provider such that it can talk to a mocked-out version of the SAP SuccessFactors
+        API, and ensure that the data it gets that way gets passed to the registration form.
+        """
+        self._configure_testshib_provider(
+            identity_provider_type='sap_success_factors',
+            other_settings=json.dumps({
+                'sapsf_oauth_root_url': 'http://successfactors.com/oauth/',
+                'sapsf_private_key': 'fake_private_key_here',
+                'odata_api_root_url': 'http://api.successfactors.com/odata/v2/',
+                'odata_company_id': 'NCC1701D',
+                'odata_client_id': 'TatVotSEiCMteSNWtSOnLanCtBGwNhGB',
+            })
+        )
+        super(SuccessFactorsIntegrationTest, self).test_register()
+
+    def test_register_http_failure(self):
+        """
+        Ensure that if there's an HTTP failure while fetching metadata, we continue, using the
+        metadata from the SAML assertion.
+        """
+        self._configure_testshib_provider(
+            identity_provider_type='sap_success_factors',
+            other_settings=json.dumps({
+                'sapsf_oauth_root_url': 'http://successfactors.com/oauth-fake/',
+                'sapsf_private_key': 'fake_private_key_here',
+                'odata_api_root_url': 'http://api.successfactors.com/odata/v2/',
+                'odata_company_id': 'NCC1701D',
+                'odata_client_id': 'TatVotSEiCMteSNWtSOnLanCtBGwNhGB',
+            })
+        )
+        self.USER_EMAIL = "myself@testshib.org"
+        self.USER_NAME = "Me Myself And I"
+        self.USER_USERNAME = "myself"
+        super(SuccessFactorsIntegrationTest, self).test_register()
+
+    @skip('Test not necessar for this subclass')
```

Comment:
:D I gotta say though, I *do* like "necessar"... It sounds like the name of some type of alien overlord in a science fiction movie.